### PR TITLE
Casminst 6411 2

### PR DESCRIPTION
**Pull request by rsjostrand-hpe**
# Summary and Scope

# Risks and Mitigations

### CODE REVIEW

File: internal/domain/blade.go [hunk 1]
```diff
@@ -94,22 +94,39 @@ func (d *Domain) AddBlade(ctx context.Context, deviceTypeSlug string, cabinetOrd
 		{HardwareType: hardwaretypes.NodeBlade, Ordinal: bladeOrdinal},
 	}
 
-	exists, err = bladePath.Exists(d.datastore)
-	if err != nil {
+	//
+	// TODO EVERYTHING BELOW IS GENERIC CODE THAT SHOULD BE SHARED
+	//
+
+	var existingDescentHardware []inventory.Hardware
```
Comment (jacobsalmela) [inline internal/domain/blade.go:101]: Descendant? Noun vs. adjective 🤷

File: internal/inventory/buildout.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	}
+
+	queue := []HardwareBuildOut{
+		{
+			ID:             topLevelHardwareID,
+			ParentID:       opts.ParentHardware.ID,
+			DeviceTypeSlug: opts.DeviceTypeSlug,
+			DeviceOrdinal:  opts.DeviceOrdinal,
+
+			LocationPath: opts.ParentHardware.LocationPath, // The loop below will add on the require location token for this devices path.
```
Comment (jacobsalmela) [inline internal/inventory/buildout.go:102]: `s/require/required/`

File: internal/inventory/buildout.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+			current.ID = existingHardware.ID
+			current.ExistingHardware = &existingHardware
+		}
+
+		for _, deviceBay := range currentDeviceType.DeviceBays {
+			log.Debug().Msgf("  Device bay: %s", deviceBay.Name)
+			if deviceBay.Default != nil {
+				log.Debug().Msgf("    Default: %s", deviceBay.Default.Slug)
+
+				// Extract the ordinal
```
Comment (jacobsalmela) [inline internal/inventory/buildout.go:136]: I changed this in my branch, but I can fix it when I do the rebase of my blade sugestions

File: internal/provider/csm/import.go [hunk 4]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+		chassisXname := nodeBladeXname.Parent()
+
+		// Retrieve the chassis
+		chassisLP, err := FromXname(chassisXname)
+		if err != nil {
+			return errors.Join(fmt.Errorf("failed to determine location path for chassis %s", chassisXname), err)
+		}
+		cChassis, err := tempDatastore.GetAtLocation(chassisLP)
+		if errors.Is(err, inventory.ErrHardwareNotFound) {
+			return errors.Join(fmt.Errorf("chassis %v does not exist in temporary datastore", chassisLP), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:602]: I have been trying to keep the hardware names consistent by using `hardwaretypes.Chassis`, `hardwaretypes.NodeBlade`, etc.  Straightforward to implement when you are already fmt f'ing.  I think errors complains about capital letters, but maybe not if it is passed in as an arg.

File: internal/provider/csm/import.go [hunk 5]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+			// Create empty node blade in CANI
+			cNodeBlade = inventory.Hardware{
+				Parent:          cChassis.ID,
+				Type:            hardwaretypes.NodeBlade,
+				Status:          inventory.HardwareStatusEmpty,
+				LocationOrdinal: pointers.IntPtr(nodeBladeXname.ComputeModule),
+			}
+
+			if err := tempDatastore.Add(&cNodeBlade); err != nil {
+				return errors.Join(fmt.Errorf("failed to create empty node blade for %v in datastore", nodeBladeXname), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:624]: `%s", hardwaretypes.NodeBlade`

File: internal/provider/csm/import.go [hunk 6]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+			}
+
+		} else if err != nil {
+			return errors.Join(fmt.Errorf("failed to retrieve node blade %v from temporary datastore", nodeBladeLP), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:628]: `%s", hardwaretypes.NodeBlade`

File: internal/provider/csm/import.go [hunk 7]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+		if errors.Is(err, inventory.ErrHardwareNotFound) {
+			// Create empty node card in CANI
+			cNodeCard = inventory.Hardware{
+				Parent:          cNodeBlade.ID,
+				Type:            hardwaretypes.NodeCard,
+				Status:          inventory.HardwareStatusEmpty,
+				LocationOrdinal: pointers.IntPtr(nodeBMCXname.NodeBMC),
+			}
+			if err := tempDatastore.Add(&cNodeCard); err != nil {
+				return errors.Join(fmt.Errorf("failed to create empty node card for %v in datastore", nodeBladeXname), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:647]: `%s", hardwaretypes.NodeCard`

File: internal/provider/csm/import.go [hunk 8]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+			}
+		} else if err != nil {
+			return errors.Join(fmt.Errorf("failed to retrieve node card %v from temporary datastore", nodeBladeLP), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:650]: `%s", hardwaretypes.NodeCard`

File: internal/provider/csm/import.go [hunk 9]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+		if errors.Is(err, inventory.ErrHardwareNotFound) {
+			// Create empty node card in CANI
+			cNode = inventory.Hardware{
+				Parent:          cNodeCard.ID,
+				Type:            hardwaretypes.Node,
+				Status:          inventory.HardwareStatusEmpty,
+				LocationOrdinal: pointers.IntPtr(nodeXname.Node),
+			}
+			if err := tempDatastore.Add(&cNode); err != nil {
+				return errors.Join(fmt.Errorf("failed to create empty node for %v in datastore", nodeXname), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:669]: `%s", hardwaretypes.Node`

File: internal/provider/csm/import.go [hunk 10]
```diff
@@ -573,7 +572,108 @@ func (csm *CSM) Import(ctx context.Context, datastore inventory.Datastore) error
...
+			}
+		} else if err != nil {
+			return errors.Join(fmt.Errorf("failed to retrieve node card %v from temporary datastore", nodeBladeLP), err)
```
Comment (jacobsalmela) [inline internal/provider/csm/import.go:672]: `%s", hardwaretypes.NodeCard`

**Review by jacobsalmela — CHANGES_REQUESTED** [review]

File: internal/domain/blade.go (hunk 1 above)
Comment (rsjostrand-hpe) [reply internal/domain/blade.go]: Switch to descendant

**Review by jacobsalmela — APPROVED** [review]

**Review by shunr-hpe — APPROVED** [review]

**Pull request merged**